Sources/Datadog/Tracing/Integrations/TracingWithRUMIntegration.swift

```diff
@@ -0,0 +1,30 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+/// Provides the current RUM context tags for produced `Spans`.
+internal final class TracingWithRUMIntegration {
+    /// The RUM attributes that should be added as Span tags.
+    ///
+    /// These attributes are synchronized using a fair and recursive lock.
+    var attribues: [String: Encodable]? {
+        get { synchronize { _attribues } }
+        set { synchronize { _attribues = newValue } }
+    }
+
+    /// Fair and recursive lock.
+    private let lock = NSRecursiveLock()
+
+    /// Unsafe attributes.
+    private var _attribues: [String: Encodable]?
+
+    private func synchronize<T>(_ block: () -> T) -> T {
+        lock.lock()
+        defer { lock.unlock() }
+        return block()
+    }
```

Comment:
Do we know how `NSLock` perf compares to `DispatchQueue`? Do we have plans on wider use of _locks_ vs _queues_ for other products in similar use case (decorating one Feature events with other Feature context)?

Here we will be running this critical section for each `tracer.startSpan()` (on **user thread**), so overhead is important. If we follow this pattern, similar thing will happen in other Features.

Reply:
Yes, here a benchmark: https://www.vadimbulavin.com/benchmarking-locking-apis/

For read operation: `DispatchQueue` is 7-8 times slower than locks.
![](https://www.vadimbulavin.com/assets/images/posts/benchmarking-locking-apis/figure-1.png)

For write operation: `DispatchQueue` is 3-4 times slower than locks
![](https://www.vadimbulavin.com/assets/images/posts/benchmarking-locking-apis/figure-4.png)

`DispatchQueue` performs better when using concurrent queue tho. But prior to this change, the `Tracer` was doing a `sync` on the (serial) queue of the `RUMMonitor` from the **user thread** ([here](https://github.com/DataDog/dd-sdk-ios/blob/develop/Sources/Datadog/Tracer.swift#L203)).

> Do we have plans on wider use of locks vs queues for other products in similar use case (decorating one Feature events with other Feature context)?

No, it's on a case by case. Here I'm using a lock to avoid syncing the `Tracer\queue` from the user-thread. But it really depends on the Feature pipeline.